tests/Tests/Protocols/Wishbone.hs

```diff
@@ -0,0 +1,160 @@
+
+smallInt :: Range Int
+smallInt = Range.linear 0 10
+
+genSmallInt :: Gen Int
+genSmallInt = Gen.integral smallInt
+
+genData :: Gen a -> Gen [a]
+genData = Gen.list (Range.linear 0 150)
+
+genWishboneTransfer
+  :: (KnownNat addressWidth)
+  => Gen a
+  -> Gen (WishboneMasterRequest addressWidth a)
+genWishboneTransfer genA = Gen.choice
+  [ Read <$> genDefinedBitVector
+  , Write <$> genDefinedBitVector <*> genA ]
+
+--
+-- 'id' circuit
+--
+
+idWriteWbSt
+  :: forall a dom addressWidth
+  . (BitPack a, NFDataX a)
+  => Circuit (Wishbone dom 'Standard addressWidth a) ()
+idWriteWbSt = Circuit go
+  where
+    go (m2s, ()) = (reply <$> m2s, ())
+
+    reply WishboneM2S{..}
+      | busCycle && strobe && writeEnable =
+          (emptyWishboneS2M @a) { acknowledge = True
+                                , readData = writeData }
+      | busCycle && strobe                = emptyWishboneS2M { acknowledge = True }
+      | otherwise                         = emptyWishboneS2M
+
+idWriteStModel
+  :: (NFData a, Eq a, ShowX a)
+  => WishboneMasterRequest addressWidth a
+  -> WishboneS2M a
+  -> () -- ^ pure state
+  -> Either String ()
+idWriteStModel (Read _)    s@WishboneS2M{..} ()
+  | acknowledge && isLeft (hasX readData) = Right ()
+  | otherwise                             =
+      Left $ "Read should have been acknowledged with no DAT " <> showX s
+idWriteStModel (Write _ a) s@WishboneS2M{..} ()
+  | acknowledge && hasX readData == Right a = Right ()
+  | otherwise                               =
+      Left $ "Write should have been acknowledged with write-data as DAT " <> showX s
+
+prop_idWriteSt_model :: Property
+prop_idWriteSt_model =
+  wishbonePropWithModel @System
+    defExpectOptions
+    idWriteStModel
+    idWriteWbSt
+    (genData $ genWishboneTransfer @10 genSmallInt)
+    ()
+
+--
+-- memory element circuit
+--
+
+memoryWb
+  :: forall ramSize dom a addressWidth
+  . ( BitPack a, NFDataX a, 1 <= ramSize, KnownDomain dom, KnownNat addressWidth
+    , HiddenClockResetEnable dom, KnownNat ramSize )
+  => Circuit (Wishbone dom 'Standard addressWidth a) ()
+memoryWb = Circuit go
+  where
+    go (m2s, ()) = (reply m2s, ())
+
+    reply
+      :: Signal dom (WishboneM2S addressWidth (BitSize a `DivRU` 8) a)
+      -> Signal dom (WishboneS2M a)
+    reply request = do
+      ack' <- ack .&&. (strobe <$> request) .&&. (busCycle <$> request)
+      val <- readValue
+      pure $ (emptyWishboneS2M @a) { acknowledge = ack', readData = val }
+      where
+        read' = addr <$> request
+        writeData' = writeData <$> request
+        write = mux ((\WishboneM2S{..} -> writeEnable && strobe && busCycle) <$> request)
+                    (Just <$> ((,) <$> read' <*> writeData'))
+                    (pure Nothing)
+
+        readValue = blockRamU ClearOnReset (SNat @ramSize) (const undefined) read' write
+        ack = register False $ (strobe <$> request) .&&. (busCycle <$> request)
```

Comment:
On the branch for the common-sense-validation I had to rework this circuit slightly and that already addressed this delayed ACK for writes.

I can add a comment mentioning this, but as far as non-registered-feedback Standard mode goes, this is the maximum data rate already as `blockRam` has a one cycle delay. Only registered feedback would yield higher data rates. Pipelined mode would cut down on one wait cycle I think, so it would be faster in that case. I'll add a comment with all this! :+1:

Reply:
Added a doc comment mentioning the data rate and the circuit itself was updated too and now has separate acks for read and writes.